scheduler/reconcile.go

```diff
@@ -460,31 +476,52 @@ func (a *allocReconciler) computeGroup(groupName string, all allocSet) bool {
 
 		// Validate and add reconnecting allocations to the plan so they are
 		// logged.
-		a.computeReconnecting(reconnect)
-
-		// The rest of the reconnecting allocations is now untainted and will
-		// be further reconciled below.
-		untainted = untainted.union(reconnect)
+		if len(reconnect) > 0 {
+			a.computeReconnecting(reconnect)
+			// The rest of the reconnecting allocations is now untainted and will
+			// be further reconciled below.
+			untainted = untainted.union(reconnect)
+		}
 	}
 
+	if len(expiring) > 0 {
+		if tg.AvoidRescheduleOnLost {
+			untainted = untainted.union(expiring)
+		} else {
+			lost = lost.union(expiring)
+		}
+	}
 	// Determine what set of disconnecting allocations need to be rescheduled now,
 	// which ones later and which ones can't be rescheduled at all.
 	timeoutLaterEvals := map[string]string{}
 	if len(disconnecting) > 0 {
-		untaintedDisconnecting, rescheduleDisconnecting, laterDisconnecting := disconnecting.filterByRescheduleable(a.batch, true, a.now, a.evalID, a.deployment)
-
-		rescheduleNow = rescheduleNow.union(rescheduleDisconnecting)
-		untainted = untainted.union(untaintedDisconnecting)
-		rescheduleLater = append(rescheduleLater, laterDisconnecting...)
+		// If MaxClientDisconnect is enabled as well as tg.AvoidRescheduleOnLost,
+		// the reschedule policy won't be enable and the lost allocations
+		// wont be rescheduled, there is no need to specifically do any extra processing.
+		// If MaxClientDisconnect is not enabled, then AvoidRescheduleOnLost
+		// requires handling.
```

Comment:
I've been debating myself over that one, `AvoidRescheduleOnLost` is "ignored here because it is already taken care of by the disabled reschedule policy so techically it is not necessary to check here. I could add it and it would make it explicit in the code, but the end result would be the same. I opted for simplicity, but conceptually, they are not mutually exclusive.